### PR TITLE
Add OPT Backbone and Tokenizer

### PR DESCRIPTION
**Pull request by mattdangerw**
I'm pretty sure this is the rough model structure. Sadly this is not quite 1-1 with the BART decoder as claimed, OPT will project from a separate `embedding_dim` to a `hidden_dim`, somewhat similar to ALBERT.

Will leave this as a draft until I have finished checkpoint conversion, in case a few more changes are needed!

### CODE REVIEW

File: keras_nlp/models/opt/opt_backbone.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+    Args:
+        vocabulary_size: int. The size of the token vocabulary.
+        num_layers: int. The number of transformer encoder layers and
+            transformer decoder layers.
+        num_heads: int. The number of attention heads for each transformer.
+            The hidden size must be divisible by the number of attention heads.
+        embedding_dim: int. The size of both the input token embedding and
+            the output sequence representation.
+        hidden_dim: int. The size of the transformer encoding and pooler layers.
```
Comment (jbischof) [inline keras_nlp/models/opt/opt_backbone.py:51]: transformer decoding?

File: keras_nlp/models/opt/opt_backbone.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    def __init__(
+        self,
+        vocabulary_size,
+        num_layers,
+        num_heads,
+        embedding_dim,
+        hidden_dim,
+        intermediate_dim,
+        dropout=0.1,
+        max_sequence_length=1024,
```
Comment (jbischof) [inline keras_nlp/models/opt/opt_backbone.py:90]: I thought the OPT sequence length was twice as long?

File: keras_nlp/models/opt/opt_backbone.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        )
+
+        # Embed tokens, potentially at a lower dimensionality.
+        token_embedding = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=embedding_dim,
+            embeddings_initializer=opt_kernel_initializer(),
+            name="token_embedding",
+        )(token_ids)
+        # Project tokens to the hidden_dim in necessary.
```
Comment (jbischof) [inline keras_nlp/models/opt/opt_backbone.py:106]: I noticed for BART that @abheesht17 did this even though the dimension already matched. You'll have to look at the checkpoints to see what's needed.

File: keras_nlp/models/opt/opt_backbone.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                intermediate_dim=intermediate_dim,
+                num_heads=num_heads,
+                dropout=dropout,
+                activation=lambda x: keras.activations.gelu(
+                    x, approximate=False
+                ),
+                layer_norm_epsilon=1e-5,
+                kernel_initializer=opt_kernel_initializer(),
+                name=f"transformer_layer_{i}",
+                has_cross_attention=True,
```
Comment (jbischof) [inline keras_nlp/models/opt/opt_backbone.py:135]: Why is this needed for a decoder-only model?

**Review by jbischof — COMMENTED** [review]
Great start! Do we not need a preprocessor or is that just WIP?

File: keras_nlp/models/opt/opt_backbone.py (hunk 3 above)
Comment (abheesht17) [reply keras_nlp/models/opt/opt_backbone.py:106]: Yep! I think I did it for FNet, and not BART.

In the original code for FNet, they actually do project it to "d_model" (from "d_emb"). However, "d_model" is always equal to "d_emb" in the configs for the pretrained models.

https://github.com/google-research/google-research/blob/master/f_net/layers.py#L387

Comment (mattdangerw) [reply keras_nlp/models/opt/opt_backbone.py:106]: Yep! I check, it is not the same for all opt checkpoints. We need this, in fact, for the very first 300M parameter checkpoint I was going to add.

File: keras_nlp/models/opt/opt_backbone.py (hunk 4 above)
Comment (mattdangerw) [reply keras_nlp/models/opt/opt_backbone.py:135]: just a copy pasta error :)

**Comment by mattdangerw** [comment]
> Great start! Do we not need a preprocessor or is that just WIP?

Preprocessor may not be a useful as the raw tokenizer for the type of pretraining we are interested in doing, so I was punting on it for now. We should add it as a follow up, but not in this PR!

File: keras_nlp/models/opt/opt_backbone.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/models/opt/opt_backbone.py:90]: You right! Will update this. Still halfway through converting a checkpoint.

**Comment by mattdangerw** [comment]
OK! Few updates on this.

First, the 350m checkpoint is an outlier. It is the only checkpoint with a projection from the token embedding to the hidden dim (and back), it applies layernorm in a different place then other checkpoints, and it has no final layernorm. Rather than handle these discrepancies, and I will just ignore the 350m checkpoint.

Second, I will update this to stay in sync with https://github.com/keras-team/keras-nlp/pull/704

With that, this should be now ready for review.

File: keras_nlp/models/opt/opt_backbone.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        )(token_ids)
+
+        # Apply successive transformer decoder blocks.
+        for i in range(num_layers):
+            x = TransformerDecoder(
+                intermediate_dim=intermediate_dim,
+                num_heads=num_heads,
+                dropout=dropout,
+                activation="relu",
+                layer_norm_epsilon=1e-5,
```
Comment (jbischof) [inline keras_nlp/models/opt/opt_backbone.py:115]: Were you able to verify all the constants in the original repo? I know checkpoint conversion doesn't catch this.

File: keras_nlp/models/opt/opt_tokenizer.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+class OPTTokenizer(BytePairTokenizer):
+    """An OPT tokenizer using Byte-Pair Encoding subword segmentation.
+
+    This tokenizer class will tokenize raw strings into integer sequences and
+    is based on `keras_nlp.tokenizers.BytePairTokenizer`. Unlike the
+    underlying tokenizer, it will check for all special tokens needed by OPT
+    models and provides a `from_preset()` method to automatically download
+    a matching vocabulary for a OPT preset.
+
+    This tokenizer does not provide truncation or padding of inputs.
```
Comment (jbischof) [inline keras_nlp/models/opt/opt_tokenizer.py:32]: Is the a BPE thing or OPT thing?

File: keras_nlp/models/opt/opt_tokenizer.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        vocabulary: string or dict, maps token to integer ids. If it is a
+            string, it should be the file path to a json file.
+        merges: string or list, contains the merge rule. If it is a string,
+            it should be the file path to merge rules. The merge rule file
+            should have one merge rule per line. Every merge rule contains
+            merge entities separated by a space.
+
+    Examples:
+
+    Batched inputs.
```
Comment (jbischof) [inline keras_nlp/models/opt/opt_tokenizer.py:49]: Weren't you wanting a `from_preset` example in the docstring?

File: keras_nlp/models/opt/opt_tokenizer.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+            vocabulary=vocabulary,
+            merges=merges,
+            **kwargs,
+        )
+
+        # We use `"</s>"` as both a start and end token, as OPT was only
+        # pre-trained with `"</s>"` marking document boundaries.
+        start_token = "</s>"
+        pad_token = "<pad>"
+        end_token = "</s>"
```
Comment (jbischof) [inline keras_nlp/models/opt/opt_tokenizer.py:99]: Where is the OOV token?

**Review by jbischof — APPROVED** [review]
Looks great, but I'm confused why we wouldn't need preprocessors for our use case (esp. if the tokenization does not support truncation or padding). Seems like a followup is required before we can pre-train.

**Comment by mattdangerw** [comment]
> Looks great, but I'm confused why we wouldn't need preprocessors for our use case (esp. if the tokenization does not support truncation or padding). Seems like a followup is required before we can pre-train.

I'm down to add the preprocessor as a follow up! We should model small, incremental PRs. So throwing it all together here would be bad.

I am also worried there is a technical point that keeps getting lost though--the real preprocessing for OPT and GPT style pretraining is **not representable with a preprocessing layer**! See https://github.com/keras-team/keras-nlp/issues/701. This isn't something to be fixed up at some point, it's a fundamental incompatibility with the assumptions in the `dataset.map(preprocessing_layer)` assumption, namely that a raw sample is 1-1 mappable with an output sample.

OPT pre-training preprocessing will look something like

```
ds = load(...)
ds.map(tokenizer).map(add_eos)
ds.rebatch(sequence_length).batch(batch_size)
```

This is quite readable and natural looking a tokenizer. This couldn't possibly work with preprocessing layers unless we wanted preprocessors to output raggeds, which seems way out of scope for this PR.

File: keras_nlp/models/opt/opt_backbone.py (hunk 5 above)
Comment (mattdangerw) [reply keras_nlp/models/opt/opt_backbone.py:115]: Yep! This is correct I believe

https://github.com/facebookresearch/metaseq/blob/00d1ac70199936ebd6d42a4a0bb82f586fd7b068/metaseq/modules/layer_norm.py#L28

File: keras_nlp/models/opt/opt_tokenizer.py (hunk 6 above)
Comment (mattdangerw) [reply keras_nlp/models/opt/opt_tokenizer.py:32]: This is an all tokenizer thing! This matches our documentation for all tokenizers -> https://keras.io/api/keras_nlp/models/bert/bert_tokenizer/

File: keras_nlp/models/opt/opt_tokenizer.py (hunk 7 above)
Comment (mattdangerw) [reply keras_nlp/models/opt/opt_tokenizer.py:49]: Can add that in when we add presets? Seems weird to submit the preset code examples before the presets. Would break some `extra_large` tests too!

File: keras_nlp/models/opt/opt_tokenizer.py (hunk 8 above)
Comment (mattdangerw) [reply keras_nlp/models/opt/opt_tokenizer.py:99]: Interesting there is none! The gpt tokenizer has a vocab entry to every byte I believe, and operates at the byte level, so I don't think there is a way to produce an input string that would create an oov output.

Though for our byte tokenizer, if you make a vocabulary without every character, you can easily actually get it to output `-1` and break an embedding layer (I've been meaning to open an issue).

But code here is correct, no oov for opt!
